File: s06_polynomial_regression.R
# Polynomial Regression

# Importing the dataset
dataset = read.csv('data/s6_Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# Not done in this case, because only ten observations are available
# install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 2/3)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting Linear Regression to the dataset [§6 Lect62: "Polynomial Regression - R pt2" ...for comparison]
lin_reg = lm(formula = Salary ~ .,
             data = dataset)

# Fitting Polynomial Regression to the dataset [§6 Lect62: "Polynomial Regression - R pt2"]
dataset$Level2 = dataset$Level^2 # Add a column (called "Level2") to the dataframe, which is Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4 # Video eventually says (Lect63 17'30-18'40) to include degree = 4
poly_reg = lm(formula = Salary ~ .,
              data = dataset)

# Visualising the Linear Regression results [§6 Lect63: "Polynomial Regression - R pt3" ...for comparison]
# install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Linear Regression)') +
  xlab('Level') +
  ylab('Salary')

# Visualising the Polynomial Regression results [§6 Lect63: "Polynomial Regression - R pt3"]
# install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Polynomial Regression)') +
  xlab('Level') +
  ylab('Salary')

# Visualising the Regression Model results (add x_grid ...if we want higher resolution and smoother curve)
# [§6 Lect63: "Polynomial Regression - R pt3" at 19'30 says this will be in the file (but not in video)]
# install.packages('ggplot2')
library(ggplot2)
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = x_grid, y = predict(poly_reg,
                                        newdata = data.frame(Level = x_grid,
                                                             Level2 = x_grid^2,
                                                             Level3 = x_grid^3,
                                                             Level4 = x_grid^4))),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Polynomial Regression)') +
  xlab('Level') +
  ylab('Salary')

# Predicting a new result with Linear Regression [§6 Lect64: "Polynomial Regression - R pt4" ...for comparison]
y_pred = predict(lin_reg, data.frame(Level = 6.5)) # This is now a single prediction, not a vector

# Predicting a new result with Polynomial Regression [§6 Lect64: "Polynomial Regression - R pt4"]
y_pred = predict(poly_reg, data.frame(Level = 6.5,
                             Level2 = 6.5^2,
                             Level3 = 6.5^3,
                             Level4 = 6.5^4))

File: s06_polynomial_regression.py
# Polynomial Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('data/s6_Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set [NB. `sklearn.cross_validation` is deprecated]
# Not done in this case, because only ten observations are available
"""from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)"""

# Fitting Linear Regression to the dataset [§6 Lect57: "Polynomial Regression - Python pt2" ...for comparison]
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Fitting Polynomial Regression to the dataset [§6 Lect57: "Polynomial Regression - Python pt2"]
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4) # Video eventually says (~15'00-17'10) to use degree = 4, not degree = 2
X_poly = poly_reg.fit_transform(X)
# cont'd : from 9min20 : include this fit (that we made with our `poly_reg` object 
# and our `X_poly` matrix of polynomial features) into a MLR model
# poly_reg.fit(X_poly, y)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

# Visualising the Linear Regression results [§6 Lect58: "Polynomial Regression - Python pt3" ...for comparison]
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Visualising the Polynomial Regression results [§6 Lect58: "Polynomial Regression - Python pt3"]
plt.scatter(X, y, color = 'red')
# plt.plot(X, lin_reg_2.predict(X_poly), color = 'blue')
plt.plot(X, lin_reg_2.predict(poly_reg.fit_transform(X)), color = 'blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Visualising the Polynomial Regression results (add X_grid ...if we want higher resolution and smoother curve)
# [§6 Lect58: "Polynomial Regression - Python pt3" from 18min00]
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1)) # to turn a vector into a matrix
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title('Truth or Bluff (Polynomial Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Predicting a new result with Linear Regression [§6 Lect59: "Polynomial Regression - Python pt4" ...for comparison]
lin_reg.predict(6.5)

# Predicting a new result with Polynomial Regression [§6 Lect59: "Polynomial Regression - Python pt4"]
lin_reg_2.predict(poly_reg.fit_transform(6.5))